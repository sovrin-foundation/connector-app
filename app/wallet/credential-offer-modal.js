// @flow
import React, { PureComponent } from 'react'
import { StyleSheet, Platform } from 'react-native'
import {
  CustomView,
  Icon,
  CustomText,
  CustomButton,
  Container,
  CustomModal,
} from '../components'
import {
  OFFSET_1X,
  OFFSET_2X,
  OFFSET_3X,
  OFFSET_6X,
  OFFSET_7X,
  color,
  HAIRLINE_WIDTH,
  whiteSmoke,
  isiPhone5,
} from '../common/styles/constant'
import Modal from 'react-native-modal'
import PaymentFailureModal from './payment-failure-modal'
import {
  CLAIM_REQUEST_STATUS,
  CREDENTIAL_OFFER_MODAL_STATUS,
} from '../claim-offer/type-claim-offer'
import type {
  CredentialOfferModalProps,
  CredentialOfferModalState,
  ClaimRequestStatusModalProps,
  ClaimRequestStatus,
} from '../claim-offer/type-claim-offer'
import { ClaimRequestStatusModal } from '../claim-offer/claim-request-modal'
import { LedgerFeesModal } from '../components/ledger-fees-modal/ledger-fees-modal'

export default class CredentialOfferModal extends PureComponent<
  CredentialOfferModalProps,
  CredentialOfferModalState
> {
  onContinue = () => {
    this.props.onClose()
  }

  render() {
    const {
      testID,
      onClose,
      claimRequestStatus,
      onRetry,
      claimOfferData,
      isValid,
      logoUrl,
      payTokenValue,
      credentialOfferModalStatus,
      transferAmount,
      connectionName,
    } = this.props
    return (
      <Modal
        backdropColor={whiteSmoke}
        isVisible={
          credentialOfferModalStatus !== CREDENTIAL_OFFER_MODAL_STATUS.NONE
        }
        animationIn="zoomIn"
        animationOut="zoomOut"
        animationOutTiming={100}
        useNativeDriver={true}
        hideModalContentWhileAnimating={true}
        onBackButtonPress={this.props.onClose}
        onBackdropPress={this.props.onClose}
        onModalHide={this.props.onModalHide}
        testID={'credential-offer-modal'}
        style={[
          claimRequestStatus ===
            CLAIM_REQUEST_STATUS.SENDING_PAID_CREDENTIAL_REQUEST ||
          claimRequestStatus === CLAIM_REQUEST_STATUS.SENDING_CLAIM_REQUEST ||
          claimRequestStatus === CLAIM_REQUEST_STATUS.SEND_CLAIM_REQUEST_SUCCESS
            ? {
                padding: 0,
                margin: 0,
              }
            : {},
        ]}
      >
        {isValid &&
          credentialOfferModalStatus ===
            CREDENTIAL_OFFER_MODAL_STATUS.INSUFFICIENT_BALANCE &&
          payTokenValue && (
            <CustomView center fifth shadow style={[styles.container]}>
              <CustomView
                spaceBetween
                style={[styles.innerContainer]}
                center
                verticalSpace
              >
                <CustomView center verticalSpace>
                  <Icon src={require('../images/alertInfo.png')} />
                </CustomView>
                <CustomText
                  transparentBg
                  style={[styles.fontBlack]}
                  center
                  demiBold
                >
                  You do not have enough tokens to purchase this credential
                </CustomText>
              </CustomView>
              <CustomButton
                fifth
                onPress={this.props.onClose}
                title={'Continue'}
                textStyle={{ fontWeight: 'bold', color: '#85BF43' }}
              />
            </CustomView>
          )}

        {isValid &&
          credentialOfferModalStatus ===
            CREDENTIAL_OFFER_MODAL_STATUS.CREDENTIAL_REQUEST_FAIL && (
            <CustomView center fifth shadow style={[styles.container]}>
              <CustomView
                spaceBetween
                style={[styles.innerContainer]}
                center
                verticalSpace
              >
                <CustomView center verticalSpace>
                  <Icon src={require('../images/alertInfo.png')} />
                </CustomView>
                <CustomText
                  transparentBg
                  style={[styles.fontBlack]}
                  center
                  demiBold
                >
                  Error accepting credential. Please try again.
                </CustomText>
              </CustomView>
              <CustomButton
                fifth
                onPress={this.props.onClose}
                title={'OK'}
                textStyle={{ fontWeight: 'bold' }}
              />
            </CustomView>
          )}

        {claimRequestStatus &&
          claimOfferData &&
          isValid &&
          credentialOfferModalStatus ===
            CREDENTIAL_OFFER_MODAL_STATUS.CREDENTIAL_REQUEST_STATUS && (
            <ClaimRequestStatusModal
              claimRequestStatus={claimRequestStatus}
              payload={claimOfferData}
              onContinue={this.props.onClose}
              senderLogoUrl={logoUrl}
              payTokenValue={payTokenValue}
              message1={payTokenValue ? 'You paid' : 'You accepted'}
              message3={payTokenValue ? '' : 'from'}
              message6={
                payTokenValue ? 'They will issue it to you shortly.' : ''
              }
              isPending={
                claimRequestStatus ===
                  CLAIM_REQUEST_STATUS.SENDING_PAID_CREDENTIAL_REQUEST ||
                claimRequestStatus ===
                  CLAIM_REQUEST_STATUS.SENDING_CLAIM_REQUEST
              }
              onModalHide={this.props.onModalHide}
            />
          )}

        {credentialOfferModalStatus ===
          CREDENTIAL_OFFER_MODAL_STATUS.SEND_PAID_CREDENTIAL_REQUEST_FAIL ||
        credentialOfferModalStatus ===
          CREDENTIAL_OFFER_MODAL_STATUS.TOKEN_SENT_FAIL ? (
          <PaymentFailureModal
            connectionName={
              connectionName ||
              (claimOfferData != null && claimOfferData.issuer != null
                ? claimOfferData.issuer.name
                : '')
            }
            testID={`${testID}-payment-failure-modal`}
            onClose={this.props.onClose}
            onRetry={this.props.onRetry}
          />
        ) : null}

        {(payTokenValue || transferAmount) &&
        credentialOfferModalStatus ===
          CREDENTIAL_OFFER_MODAL_STATUS.LEDGER_FEES ? (
          <LedgerFeesModal
            onNo={this.props.onNo}
            onYes={this.props.onYes}
            transferAmount={transferAmount}
            renderFeesText={this.props.renderFeesText}
          />
        ) : null}
      </Modal>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginHorizontal: OFFSET_3X,
  },
  innerContainer: {
    ...Platform.select({
      ios: {
        borderBottomColor: color.bg.fifth.font.tertiary,
        borderBottomWidth: HAIRLINE_WIDTH,
      },
      android: {
        borderBottomColor: color.bg.fifth.font.secondary,
        borderBottomWidth: 1,
      },
    }),
    padding: OFFSET_2X,
  },
  message: {
    marginBottom: OFFSET_1X / 2,
  },
  fontBlack: {
    color: '#777777',
  },
})
