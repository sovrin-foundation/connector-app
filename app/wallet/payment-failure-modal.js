// @flow
import React, { PureComponent } from 'react'
import { StyleSheet, Platform } from 'react-native'
import {
  CustomView,
  Icon,
  CustomText,
  CustomButton,
  Container,
} from '../components'
import {
  OFFSET_1X,
  OFFSET_2X,
  OFFSET_3X,
  OFFSET_6X,
  OFFSET_7X,
  color,
  HAIRLINE_WIDTH,
  whiteSmoke,
} from '../common/styles/constant'
import Modal from 'react-native-modal'

type PaymentFailureModalProps = {
  testID: string,
  onClose: () => void,
  onRetry: (event: any) => void,
  connectionName: string,
}

export default class PaymentFailureModal extends PureComponent<
  PaymentFailureModalProps,
  void
> {
  render() {
    const { testID, onClose } = this.props
    return (
      <CustomView fifth shadow style={[styles.container]}>
        <CustomView spaceBetween style={[styles.innerContainer]}>
          <Icon
            iconStyle={[{ margin: 10 }]}
            src={require('../images/alertInfo.png')}
            extraLarge
            center
            resizeMode="contain"
            testID={`${testID}-modal-header-icon`}
          />
          <CustomText
            h5
            center
            tertiary
            bg="tertiary"
            transparentBg
            style={[styles.message]}
            demiBold
            testID={`${testID}-modal-title`}
          >
            {'Payment Failure'}
          </CustomText>
          <CustomText
            h5
            center
            tertiary
            bg="tertiary"
            transparentBg
            style={[styles.message]}
            demiBold
            testID={`${testID}-modal-content`}
          >
            {`Something went wrong trying to pay ${
              this.props.connectionName
            }. Please try again.`}
          </CustomText>
        </CustomView>
        <CustomView row>
          <Container>
            <CustomButton
              fifth
              onPress={onClose}
              title={'Cancel'}
              testID={`${testID}-modal-cancel`}
              textStyle={{ fontWeight: 'bold' }}
            />
          </Container>
          <Container>
            <CustomButton
              fifth
              onPress={this.props.onRetry}
              title={'Retry'}
              testID={`${testID}-modal-retry`}
              textStyle={{ fontWeight: 'bold' }}
            />
          </Container>
        </CustomView>
      </CustomView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginHorizontal: OFFSET_3X,
  },
  innerContainer: {
    ...Platform.select({
      ios: {
        borderBottomColor: color.bg.fifth.font.tertiary,
        borderBottomWidth: HAIRLINE_WIDTH,
      },
      android: {
        borderBottomColor: color.bg.fifth.font.secondary,
        borderBottomWidth: 1,
      },
    }),
    padding: OFFSET_2X,
  },
  message: {
    marginBottom: OFFSET_1X / 2,
  },
})
