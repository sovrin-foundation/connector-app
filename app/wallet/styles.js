// @flow
import { StyleSheet, Dimensions } from 'react-native'
import {
  PADDING_HORIZONTAL,
  PADDING_VERTICAL,
  isBiggerThanShortDevice,
  color,
} from '../common/styles/constant'

const { width } = Dimensions.get('window')
const SPACE_FILLER = 50
const amountSpacing = isBiggerThanShortDevice ? PADDING_VERTICAL * 2 : 0
const ctaButtonHeight = isBiggerThanShortDevice ? 63 : 43

const styles = StyleSheet.create({
  headerCloseIcon: {
    marginTop: PADDING_VERTICAL,
  },
  ctaButton: {
    flex: -1,
    height: ctaButtonHeight,
    borderRadius: 5,
    backgroundColor: color.bg.eighth.color,
  },
  alignItemsCenter: {
    marginBottom: 6,
    marginHorizontal: '5%',
  },
  verticalSpacing: {
    flexDirection: 'column',
    alignContent: 'center',
    justifyContent: 'space-around',
  },
})

export default styles
