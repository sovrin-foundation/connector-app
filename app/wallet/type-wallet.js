// @flow
import type { ReactNavigation } from '../common/type-common'
import type { CustomError } from '../common/type-common'
import type { IsValid } from '../components/input-control/type-input-control'
import type { CredentialOfferModalStatus } from '../claim-offer/type-claim-offer'

export type BackupWalletProps = {
  backup: BackupInfo,
  render: (backup: string, backupWallet: () => void) => void,
  walletBackup: () => void,
} & ReactNavigation

export type BackupInfo = {
  latest: ?string,
  error: ?CustomError,
  backupPath: ?string,
} & StoreStatus

export const BACKUP_WALLET = 'BACKUP_WALLET'

export const BACKUP_WALLET_FAIL = 'BACKUP_WALLET_FAIL'

export const BACKUP_WALLET_SUCCESS = 'BACKUP_WALLET_SUCCESS'

export const GET_WALLET_ENCRYPTION_KEY = 'GET_WALLET_ENCRYPTION_KEY'

export const BACKUP_WALLET_PATH = 'BACKUP_WALLET_PATH'

export const SHARE_WALLET_BACKUP = 'SHARE_WALLET_BACKUP'

export const ERROR_BACKUP_WALLET = {
  code: 'WB-001',
  message: 'Error while backing up wallet',
}

export const ERROR_BACKUP_WALLET_SHARE = {
  code: 'WB-002',
  message: 'Error while sharing zipped backup',
}

export type WalletTabSendDetailsProps = {
  tokenAmount: string,
  tokenSentStatus: $Keys<typeof STORE_STATUS>,
  sendTokens: (
    tokenAmount: string,
    recipientAddress: string
  ) => SendTokensAction,
} & ReactNavigation

export type BackupWalletAction = {
  type: typeof BACKUP_WALLET,
  data: BackupInfo,
}

export type ShareBackupAction = {
  type: typeof SHARE_WALLET_BACKUP,
  data: BackupInfo,
}

export type WalletBalanceProps = {
  walletBalance: string,
  refreshWalletBalance: () => {},
  render: Function,
}

export type WalletSendAmountState = {
  text: string,
}

export type WalletSendAmountProps = {
  screenProps: ReactNavigation,
  selectTokenAmount: (tokenAmount: string) => {},
  walletBalance: string,
  paymentStatus: $Keys<typeof STORE_STATUS>,
} & ReactNavigation

export type WalletProps = {} & ReactNavigation

export type WalletPayload = {
  walletPath: string,
  encryptionKey: string,
}

export type WalletTabReceiveProps = {
  walletAddresses: Array<string>,
  refreshWalletAddresses: () => {},
  promptBackupBanner: (showBackup: boolean) => {},
  addressStatus: $Keys<typeof STORE_STATUS>,
  currentScreen: string,
}

export type WalletTabReceiveState = {
  copyButtonText: string,
}

export type WalletSendPaymentData = {
  paymentTo: string,
  paymentFor?: string,
}

export type WalletTabSendDetailsState = {
  showPaymentAddress: boolean,
  isPaymentAddressValid: IsValid,
  credentialOfferModalStatus: CredentialOfferModalStatus,
}

export type WalletHistoryProps = {
  walletHistory: WalletHistory,
  refreshWalletHistory: () => {},
}

export type WalletTabsProps = {} & ReactNavigation

export type WalletHistoryEvent = {
  id: string,
  senderName?: string,
  senderAddress: string,
  action: string,
  tokenAmount: string,
  timeStamp: string,
}

export type WalletBalance = {
  data: string,
} & StoreStatus &
  StoreError

export type WalletAddresses = {
  data: Array<string>,
} & StoreStatus &
  StoreError

export type WalletHistory = {
  transactions: WalletHistoryEvent[],
} & StoreStatus &
  StoreError

export type Payment = {
  tokenAmount: string,
} & StoreStatus &
  StoreError

export const ERROR_LOADING_WALLET = {
  code: 'W001',
  message: 'Error while loading wallet data',
}

export const ERROR_REFRESHING_WALLET_BALANCE = {
  code: 'W002',
  message: 'Error while refreshing wallet balance',
}

export const ERROR_REFRESHING_WALLET_ADDRESSES = {
  code: 'W003',
  message: 'Error while refreshing wallet addresses',
}

export const ERROR_REFRESHING_WALLET_HISTORY = {
  code: 'W004',
  message: 'Error while refreshing wallet history',
}

export const ERROR_LOADING_WALLET_BALANCE = {
  code: 'W005',
  message: 'Error while loading wallet balance',
}

export const ERROR_LOADING_WALLET_ADDRESSES = {
  code: 'W006',
  message: 'Error while loading wallet addresses',
}

export const ERROR_LOADING_WALLET_HISTORY = {
  code: 'W007',
  message: 'Error while loading wallet history',
}

export const ERROR_SENDING_TOKENS = {
  code: 'W008',
  message: 'Error sending tokens',
}

export const ERROR_SENDING_TOKENS_WITH_FEES = {
  code: 'W009',
  message:
    'Error sending tokens while transfer fees is greater than or equal to transfer amount',
}

export const STORE_STATUS = {
  IDLE: 'IDLE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
}

export type StoreError = { error: ?CustomError }
export type StoreStatus = { status: $Keys<typeof STORE_STATUS> }

export type WalletStore = {
  walletBalance: WalletBalance,
  walletAddresses: WalletAddresses,
  walletHistory: WalletHistory,
  backup: BackupInfo,
  payment: Payment,
}

export type HydrateWalletBalanceFailAction = {
  type: typeof HYDRATE_WALLET_BALANCE_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type HydrateWalletAddressesFailAction = {
  type: typeof HYDRATE_WALLET_ADDRESSES_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type HydrateWalletHistoryFailAction = {
  type: typeof HYDRATE_WALLET_HISTORY_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type SendTokensAction = {
  type: typeof SEND_TOKENS,
  tokenAmount: string,
  recipientWalletAddress: string,
}

export type SelectTokensAction = {
  type: typeof SEND_TOKENS,
  tokenAmount: string,
}

export const HYDRATE_WALLET_BALANCE_FAIL = 'HYDRATE_WALLET_BALANCE_FAIL'
export const HYDRATE_WALLET_ADDRESSES_FAIL = 'HYDRATE_WALLET_ADDRESSES_FAIL'
export const HYDRATE_WALLET_HISTORY_FAIL = 'HYDRATE_WALLET_HISTORY_FAIL'
export const REFRESH_WALLET_BALANCE_FAIL = 'REFRESH_WALLET_BALANCE_FAIL'
export const REFRESH_WALLET_ADDRESSES_FAIL = 'REFRESH_WALLET_ADDRESSES_FAIL'
export const REFRESH_WALLET_HISTORY_FAIL = 'REFRESH_WALLET_HISTORY_FAIL'
export const HYDRATE_WALLET_BALANCE = 'HYDRATE_WALLET_BALANCE'
export const HYDRATE_WALLET_ADDRESSES = 'HYDRATE_WALLET_ADDRESSES'
export const HYDRATE_WALLET_HISTORY = 'HYDRATE_WALLET_HISTORY'
export const REFRESH_WALLET = 'REFRESH_WALLET'
export const REFRESH_WALLET_ADDRESSES = 'REFRESH_WALLET_ADDRESSES'
export const REFRESH_WALLET_HISTORY = 'REFRESH_WALLET_HISTORY'
export const REFRESH_WALLET_BALANCE = 'REFRESH_WALLET_BALANCE'
export const WALLET_BALANCE_REFRESHED = 'WALLET_BALANCE_REFRESHED'
export const WALLET_ADDRESSES_REFRESHED = 'WALLET_ADDRESSES_REFRESHED'
export const WALLET_HISTORY_REFRESHED = 'WALLET_HISTORY_REFRESHED'
export const SEND_TOKENS = 'SEND_TOKENS'
export const SEND_TOKENS_FAIL = 'SEND_TOKENS_FAIL'
export const TOKEN_SENT_SUCCESS = 'TOKEN_SENT_SUCCESS'
export const SELECT_TOKEN_AMOUNT = 'SELECT_TOKEN_AMOUNT'
export const PROMPT_WALLET_BACKUP_BANNER = 'PROMPT_WALLET_BACKUP_BANNER'
export const WALLET_ADDRESSES_FETCH_START = 'WALLET_ADDRESSES_FETCH_START'

export type PromptBackupBannerAction = {
  type: typeof PROMPT_WALLET_BACKUP_BANNER,
  showBanner: boolean,
}

export type HydrateWalletBalanceData = string

export type HydrateWalletAddressesData = Array<string>

export type HydrateWalletHistoryTransactions = WalletHistoryEvent[]

export type RefreshWalletHistoryAction = {
  type: typeof REFRESH_WALLET_HISTORY,
}

export type WalletBalanceRefreshedAction = {
  type: typeof WALLET_BALANCE_REFRESHED,
  walletBalance: WalletBalance,
}

export type WalletAddressesRefreshedAction = {
  type: typeof WALLET_ADDRESSES_REFRESHED,
  walletAddresses: WalletAddresses,
}

export type WalletHistoryRefreshedAction = {
  type: typeof WALLET_HISTORY_REFRESHED,
  walletHistory: WalletHistory,
}

export type RefreshWalletAddressesAction = {
  type: typeof REFRESH_WALLET_ADDRESSES,
}

export type RefreshWalletBalanceAction = {
  type: typeof REFRESH_WALLET_BALANCE,
  walletBalance: WalletBalance,
}

export type HydrateWalletBalanceAction = {
  type: typeof HYDRATE_WALLET_BALANCE,
  walletBalance: WalletBalance,
}

export type HydrateWalletAddressesAction = {
  type: typeof HYDRATE_WALLET_ADDRESSES,
  walletAddresses: WalletAddresses,
}

export type HydrateWalletHistoryAction = {
  type: typeof HYDRATE_WALLET_HISTORY,
  walletHistory: WalletHistory,
}

export type TokenSentSuccessAction = {
  type: typeof TOKEN_SENT_SUCCESS,
  payment: Payment,
}

export type WalletStoreAction =
  | HydrateWalletBalanceFailAction
  | HydrateWalletAddressesFailAction
  | HydrateWalletHistoryFailAction
  | RefreshWalletBalanceAction
  | RefreshWalletAddressesAction
  | RefreshWalletHistoryAction
  | WalletHistoryRefreshedAction
  | WalletBalanceRefreshedAction
  | WalletAddressesRefreshedAction
  | HydrateWalletBalanceAction
  | HydrateWalletAddressesAction
  | HydrateWalletHistoryAction
  | BackupWalletAction
  | SendTokensAction
  | TokenSentSuccessAction
  | SelectTokensAction
