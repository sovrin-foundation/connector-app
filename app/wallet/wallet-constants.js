// @flow

export const RECEIVE_TAB = 'receive'
export const SEND_TAB = 'send'
export const HISTORY_TAB = 'history'
export const RECEIVE_TAB_TEST_ID = 'wallet-tab-receive'
export const SEND_TAB_TEST_ID = 'wallet-tab-send'
export const HISTORY_TAB_TEST_ID = 'wallet-tab-history'
export const SEND_TOKEN_BUTTON = 'wallet-send-token-button'
export const SEND_TOKENS_TO_PAYMENT_ADDRESS = 'wallet-send-tokens-to-address'
export const TO_SEND_DETAILS_TEST_ID = 'to-send-details-test-id'
export const FOR_SEND_DETAILS_TEST_ID = 'for-send-details-test-id'

export const FONT_SIZE_MAPPING = (length: number) => {
  switch (true) {
    case length < 5:
      return 60
    case length < 7:
      return 60
    case length < 10:
      return 50
    case length < 12:
      return 40
    default:
      return 30
  }
}
