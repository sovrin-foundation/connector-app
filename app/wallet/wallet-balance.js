// @flow

import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { View, Image } from 'react-native'
import styles from './styles'
import type { WalletBalanceProps } from './type-wallet'
import { getWalletBalance } from '../store/store-selector'
import type { Store } from '../store/type-store'
import { refreshWalletBalance } from './wallet-store'

export class WalletBalance extends PureComponent<WalletBalanceProps, void> {
  componentDidMount() {
    this.props.refreshWalletBalance()
  }

  render() {
    const { walletBalance, render }: WalletBalanceProps = this.props
    return <View>{render(walletBalance)}</View>
  }
}

const mapStateToProps = (state: Store) => {
  return {
    walletBalance: getWalletBalance(state),
  }
}

const mapDispatchToProps = dispatch =>
  bindActionCreators({ refreshWalletBalance }, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(WalletBalance)
