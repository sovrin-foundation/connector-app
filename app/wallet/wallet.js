// @flow

import React, { PureComponent } from 'react'
import { Image, StyleSheet } from 'react-native'
import { createStackNavigator } from 'react-navigation'
import { scale } from 'react-native-size-matters'
import {
  Container,
  CustomView,
  CustomText,
  Icon,
  CustomHeader,
} from '../components'
import { walletRoute } from '../common'
import {
  color,
  isSmallWidthDevice,
  responsiveHorizontalPadding,
} from '../common/styles/constant'
import WalletBalance from './wallet-balance'
import WalletTabs from './wallet-tabs'
import type { WalletProps } from './type-wallet'
import styles from './styles'

const closeImage = require('../images/iconClose.png')
const sovrinLogo = require('../images/sovrinLogo.png')
const tokenLogo = require('../images/sovrinTokenWhite.png')

const walletTabsHeaderBalanceFontSize = (balance: string) => {
  /**
   * TODO:KS below logic to select size as per length is an ugly logic
   * adjustsFontSizeToFit is not working due to center alignment and no width
   * on parent container, we can't set width as well from Dimensions module
   * because there are other elements besides this which are taking up space
   * also, other elements width are not defined, so we can't calculate width as well
   * Even if we would have used adjustsFontSizeToFit, we can't control the scale
   * with which font size is reduced on Android, a scale factor is provided for ios
   * So, for now as a hack we are providing font by ourselves.
   * Normally, we would have preferred to just wrap the text, but we can't wrap it
   * because it is a number. Also, header is hard coded with a height
   */
  const digits = balance ? balance.length : 0
  switch (true) {
    case digits < 13:
      return scale(30)
    case digits < 16:
      return scale(25)
    case digits < 20:
      return scale(19)
    default:
      return scale(16)
  }
}

const walletTabsHeaderStyles = StyleSheet.create({
  outerContainer: {
    height: 120,
    paddingHorizontal: responsiveHorizontalPadding,
  },
  closeButtonContainer: { height: '100%', alignItems: 'flex-end', zIndex: 1 },
})
// sovrin currency logo beside the amount also needs to be scaled down
// this cannot be scaled down just one the basis of device
// because if wallet balance is just 5 or 6 characters, then font size
// for wallet balance will be big, so we want to scale this logo as well
// as per the size of wallet balance text as well
const sovrinTokenIconSize = (balanceLength: number) =>
  balanceLength > 14 && isSmallWidthDevice ? { small: true } : { medium: true }

export class Wallet extends PureComponent<WalletProps, void> {
  static navigationOptions = ({ navigation }) => ({
    // Makes it so that headerTitle is centered Android
    header: (
      <CustomHeader
        backgroundColor={color.actions.font.seventh}
        outerContainerStyles={walletTabsHeaderStyles.outerContainer}
        largeHeader
      >
        <Container />

        <WalletBalance
          render={(balance: string) => (
            <Container>
              <Container center>
                <Icon extraLarge src={sovrinLogo} />
              </Container>
              <CustomView row center>
                <Icon
                  {...sovrinTokenIconSize(balance ? balance.length : 0)}
                  src={tokenLogo}
                />
                <CustomView horizontalSpace>
                  <CustomText
                    medium
                    formatNumber
                    transparentBg
                    style={[
                      {
                        width: '100%',
                        fontSize: walletTabsHeaderBalanceFontSize(balance),
                      },
                    ]}
                  >
                    {balance}
                  </CustomText>
                </CustomView>
              </CustomView>
              <CustomView center>
                <CustomText h6 demiBold transparentBg uppercase>
                  current balance
                </CustomText>
              </CustomView>
            </Container>
          )}
        />

        <Container
          testID={'wallet-header-close'}
          style={[walletTabsHeaderStyles.closeButtonContainer]}
        >
          <Icon
            medium
            onPress={() => navigation.goBack(null)}
            testID="wallet-header-close-image"
            iconStyle={[styles.headerCloseIcon]}
            src={closeImage}
          />
        </Container>
      </CustomHeader>
    ),
    gesturesEnabled: true,
  })

  render() {
    return (
      <Container tertiary>
        <WalletTabs navigation={this.props.navigation} />
      </Container>
    )
  }
}

export default createStackNavigator({
  [walletRoute]: {
    screen: Wallet,
  },
})
