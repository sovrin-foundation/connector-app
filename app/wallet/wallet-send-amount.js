// @flow

import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { Text, Animated } from 'react-native'
import { scale } from 'react-native-size-matters'
import { Container, CustomView, CustomText, CustomButton } from '../components'
import { Keyboard } from '../components'
import { color, MEDIUM_DEVICE, deviceHeight } from '../common/styles/constant'
import { SEND_TOKEN_BUTTON, FONT_SIZE_MAPPING } from './wallet-constants'
import styles from './styles'
import type {
  WalletSendAmountState,
  WalletSendAmountProps,
} from './type-wallet'
import { walletTabSendDetailsRoute } from '../common/route-constants'
import { selectTokenAmount } from './wallet-store'
import { getWalletBalance } from '../store/store-selector'
import type { Store } from '../store/type-store'
import { STORE_STATUS } from './type-wallet'
import { conversionFactorLength } from '../bridge/react-native-cxs/vcx-transformers'

class WalletSendAmount extends PureComponent<
  WalletSendAmountProps,
  WalletSendAmountState
> {
  _shake: any
  _tokenKeyboard: null | Keyboard
  state = {
    text: '',
  }

  constructor(props) {
    super(props)
    this._shake = new Animated.Value(0)
  }

  changeText = (text: string, animate: boolean) => {
    if (animate) {
      return this.shake()
    }
    return this.setState({ text })
  }

  shake = () => {
    this._shake.setValue(0)
    Animated.spring(this._shake, {
      toValue: 1,
      friction: 3,
      tension: 10,
      useNativeDriver: true,
    }).start(() => {
      this._shake.setValue(0)
    })
  }

  sendTokenAmount = () => {
    this.props.selectTokenAmount(this.state.text)
    if (this.state.text.length) {
      this.props.screenProps.navigation.navigate(walletTabSendDetailsRoute, {
        ...this.props.navigation,
      })
    }
  }

  saveTokenKeyboardRef = tokenKeyboard => {
    this._tokenKeyboard = tokenKeyboard
  }

  render() {
    const { text } = this.state

    // adjust fontSize based on text length
    const fontSize = scale(FONT_SIZE_MAPPING(text.length))
    const animatedStyle = {
      transform: [
        {
          translateX: this._shake.interpolate({
            inputRange: [0, 0.1, 0.2, 0.3, 0.4, 0.55, 0.7, 0.8, 0.9, 1],
            outputRange: [0, -10, 0, -10, 0, -10, 0, -10, 0, 0],
          }),
        },
      ],
    }
    const isSendDisabled = text.length < 1 || text === '0'

    return (
      <Container tertiary style={[styles.verticalSpacing]}>
        <CustomView>
          <CustomView verticalSpace>
            <CustomText
              animated
              formatNumber
              transparentBg
              center
              style={[
                animatedStyle,
                {
                  fontSize,
                  color: color.bg.seventh.font.fifth,
                  height: deviceHeight < MEDIUM_DEVICE ? 70 : 90,
                  lineHeight: deviceHeight < MEDIUM_DEVICE ? 75 : 80,
                },
              ]}
              numberOfLines={1}
            >
              {text || 0}
            </CustomText>
          </CustomView>
          <CustomText h6 center tertiary demibold uppercase transparentBg>
            sovrin tokens
          </CustomText>
        </CustomView>
        <Keyboard
          color={color.bg.seventh.font.fifth}
          onPress={this.changeText}
          maxValue={this.props.walletBalance}
          ref={this.saveTokenKeyboardRef}
          afterDecimalSeparatorMaxLength={conversionFactorLength}
        />
        <CustomView safeArea style={[styles.alignItemsCenter]}>
          <CustomButton
            disabled={isSendDisabled}
            customColor={{ backgroundColor: color.bg.eighth.color }}
            onPress={this.sendTokenAmount}
            testID={SEND_TOKEN_BUTTON}
            style={[styles.ctaButton]}
            primary
            title="Select Recipient"
          />
        </CustomView>
      </Container>
    )
  }

  componentDidUpdate(prevProps: WalletSendAmountProps) {
    if (
      this.props.paymentStatus !== prevProps.paymentStatus &&
      this.props.paymentStatus === STORE_STATUS.SUCCESS
    ) {
      // if payment is successful, we can now reset token amount and keyboard
      this.setState({ text: '' })
      this._tokenKeyboard && this._tokenKeyboard.clear()
    }
  }
}

const mapStateToProps = (state: Store) => {
  return {
    walletBalance: getWalletBalance(state),
    paymentStatus: state.wallet.payment.status,
  }
}

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      selectTokenAmount,
    },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(WalletSendAmount)
