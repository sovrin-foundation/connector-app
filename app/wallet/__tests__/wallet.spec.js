// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { walletRoute } from '../../common'
import { Wallet } from '../wallet'
import { getNavigation, getStore } from '../../../__mocks__/static-data'

function props() {
  return {
    navigation: getNavigation(),
    route: {
      currentScreen: walletRoute,
    },
  }
}

describe('<Wallet />', () => {
  const store = getStore()

  jest.useFakeTimers()

  it('should render Wallet', () => {
    const walletProps = props()
    const wrapper = renderer
      .create(
        <Provider store={store}>
          <Wallet {...walletProps} />
        </Provider>
      )
      .toJSON()
    expect(wrapper).toMatchSnapshot()
  })
})
