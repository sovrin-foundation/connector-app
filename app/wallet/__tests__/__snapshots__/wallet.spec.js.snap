// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Wallet /> should render Wallet 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  tertiary={true}
>
  <View />
</View>
`;
